### PR TITLE
Allow UnionTypes to accept a types thunk

### PR DESCRIPTION
**Pull request by ekosz**
This should help with module loading order creating null types.

As this is my first PR for the project, please let me know if my tests/style are inline for `graphql-js`. More than happy to fix things up. Thanks!

### CODE REVIEW

**Comment by ekosz** [comment]
@leebyron Do you know what the plan is for other places in `graphql-js` that don't take a thunk?

File: src/type/__tests__/definition-test.js [hunk 1]
```diff
@@ -342,13 +342,23 @@ describe('Type System: Example', () => {
...
         `BadUnion may only contain Object types, it cannot contain: ${x}.`
       );
     });
   });
 
+  it('allows a thunk for Union\'s types', () => {
+    const types = [ ObjectType ];
+    const union = new GraphQLUnionType({
+      name: 'ThunkUnion',
+      types: () => types
```
Comment (leebyron) [inline src/type/__tests__/definition-test.js]: Isn't the whole point of using a thunk to enable a delay of referencing `ObjectType` until access time?

Shouldn't this be `types: () => [ ObjectType ]` to illustrate that usage?

**Comment by leebyron** [comment]
Thanks - there's no reason this couldn't have also accepted a thunk

File: src/type/__tests__/definition-test.js (hunk 1 above)
Comment (leebyron) [inline src/type/__tests__/definition-test.js]: Ah, I see - you're using the same reference below for equality. Makes sense!

**Comment by leebyron** [comment]
In general the idea is that thunks should be acceptable wherever breaking a circular dependency between Types might require them.

Unions are not very often in that loop since the Object type itself is typically where the cycle is found and broken, but certainly it makes sense to include them here as well.

**Pull request merged**